### PR TITLE
Windows: fixes

### PR DESCRIPTION
**Pull request by xoviat**
Attached are some fixes for the windows functions.

### CODE REVIEW

**Comment by xoviat** [comment]
I don't think there's been a single contributor pull request accepted, looking at the history.

@sscalpone Are you going to merge (not now, but eventually) this or was this a waste of time?

**Comment by sscalpone** [comment]
A few contributor requests have been merged for README.md and CMakefile.txt files.  Flang needs to have a CLA for source-file changes.  We are working to create this CLA now.  We will review on flang-dev list soon.

**Comment by xoviat** [comment]
I should mention that the timing functions don't actually work; I'm not sure what they're for.

**Comment by xoviat** [comment]
Timing issue triaged as #283. That allows this to be merged and we can fix them after we fix the higher priority of compiling "hello world."

**Comment by isuruf** [comment]
I think the issue we are seeing is the same as https://github.com/flang-compiler/flang/issues/222

**Comment by xoviat** [comment]
Yes, that's it. But it still doesn't solve the problem of `flang.exe` trying to link an object file with the `-o` option. We could try to fix that ourselves, but it will be more time efficient for @sscalpone to give us a fix, since he's probably more familiar with the C++.

**Comment by xoviat** [comment]
Note: need feedback on the `_Dcomplex` definitions. I'm not quite sure what the implications are.

**Comment by xoviat** [comment]
Somthing like:
```
extern int __fortio_binary_mode(int fd) {
    return (_setmode( _fileno( fd ), _O_BINARY ) != -1);
}
```

**Comment by xoviat** [comment]
You might or might not need `_fileno`. Not sure what the type of that struct is.

**Comment by sscalpone** [comment]
We haven't pushed out Windows implementation files to github yet.  I'm working on a pull request to get the windows sources into the repo.  With hope, they'll start showing up tomorrow (Tuesday) to support your efforts.  Do you have changes for flang-compiler/clang too?

**Comment by xoviat** [comment]
No. Changes were not required there. I implemented some of the functions in a "quick" (eg. nonworking) way just to get the compilation going. Then I triaged those functions so that I could come back to them later (#285, #283). 

@isuruf Had to fix CMake so that it would call the linker correctly.

I really wish for the CLA so that there's no duplication of work.

**Comment by xoviat** [comment]
Also, would highly recommend merging #284. We spent some time on that.

**Comment by xoviat** [comment]
Honestly, though, it feels nice to write C99. Not so much C++ (which the flang-clang) is written in. Glad I did not have to touch that.

File: runtime/flang/buffer.c [hunk 1]
```diff
@@ -72,7 +72,7 @@ __fort_zopen(char *path)
     __fort_rrecv(ioproc, &off, sizeof(off), 1, __UCHAR);
   } else {
     off = 0;
-    fd = open(path, O_WRONLY | O_CREAT | O_TRUNC, 0666);
```
Comment (xoviat) [inline runtime/flang/buffer.c]: Should be `open(path, O_WRONLY | O_CREAT | O_TRUNC, 0666);`.

File: runtime/flang/curdir.c [hunk 2]
```diff
@@ -113,4 +116,7 @@ void __fort_gethostname(host) char *host;
     p = un.nodename;
   }
   strcpy(host, p);
+#else
+  strcpy(host, "localhost");
+#endif
```
Comment (xoviat) [inline runtime/flang/curdir.c]: Should be 

```
#include <Winsock2.h>
char host[128] = "";
gethostname(host, sizeof(host));
```

**Comment by isuruf** [comment]
If anybody is interested, here's the status now.

```
[ 69%] Linking Fortran shared library ..\..\bin\flang.dll
LINK: command "C:\PROGRA~2\MICROS~1.0\VC\bin\amd64\link.exe /nologo @CMakeFiles\flang_shared.dir\objects1.rsp /out:..\..\bin\flang.dll /implib:..\..\lib\flang.lib /pdb:C:\Users\idf2\flang\build3\bin\flang.pdb /dll /version:0.0 /machine:x64 /INCREMENTAL:NO /DEF:CMakeFiles\flang_shared.dir\exports.def -LIBPATH:C:\Users\idf2\miniconda3\Library\lib ..\..\lib\flangrti.lib libomp.lib /MANIFEST /MANIFESTFILE:..\..\bin\flang.dll.manifest" failed (exit code 1120) with the following output:
etime3f.c.obj : error LNK2005: convert_filetime already defined in dtime3f.c.obj
etime3f.c.obj : error LNK2005: dtime_ already defined in dtime3f.c.obj
cvt.c.obj : warning LNK4197: export 'ecvt' specified multiple times; using first specification
cvt.c.obj : warning LNK4197: export 'fcvt' specified multiple times; using first specification
   Creating library ..\..\lib\flang.lib and object ..\..\lib\flang.exp
amod.c.obj : error LNK2019: unresolved external symbol __fmth_i_amod referenced in function __mth_i_amod
cdcos.c.obj : error LNK2019: unresolved external symbol sincos referenced in function __mth_i_cdcos
cdexp.c.obj : error LNK2001: unresolved external symbol sincos
cexp.c.obj : error LNK2019: unresolved external symbol sincosf referenced in function __mth_i_cexp
dmod.c.obj : error LNK2019: unresolved external symbol __fmth_i_dmod referenced in function __mth_i_dmod
fsync3f.c.obj : error LNK2019: unresolved external symbol fsync referenced in function fsync_
irandm3f.c.obj : error LNK2019: unresolved external symbol srand48 referenced in function irandm_
irandm3f.c.obj : error LNK2019: unresolved external symbol lrand48 referenced in function irandm_
xfer_heap_dum.c.obj : error LNK2019: unresolved external symbol sbrk referenced in function __fort_sbrk
nmlwrite.c.obj : error LNK2001: unresolved external symbol __fortio_binary_mode
rewind.c.obj : error LNK2001: unresolved external symbol __fortio_binary_mode
backspace.c.obj : error LNK2001: unresolved external symbol __fortio_binary_mode
close.c.obj : error LNK2001: unresolved external symbol __fortio_binary_mode
fmtwrite.c.obj : error LNK2001: unresolved external symbol __fortio_binary_mode
ldwrite.c.obj : error LNK2001: unresolved external symbol __fortio_binary_mode
miscsup_com.c.obj : error LNK2019: unresolved external symbol gettimeofday referenced in function fort_dandt
type.c.obj : error LNK2019: unresolved external symbol __get_fort_size_of referenced in function f90_poly_asn
utils.c.obj : error LNK2019: unresolved external symbol ftruncate referenced in function __fortio_rwinit
init.c.obj : error LNK2019: unresolved external symbol __fort_sethand referenced in function __fort_begpar
..\..\bin\flang.dll : fatal error LNK1120: 13 unresolved externals
```

**Comment by xoviat** [comment]
@isuruf did you add my `__fortio_binary_mode`?

**Comment by xoviat** [comment]
Also, `gettineofday` is at https://github.com/Microsoft/vcpkg/pull/1991.

**Comment by isuruf** [comment]
> @isuruf did you add my __fortio_binary_mode?

No, I didn't. I'll do that once I get some time.

**Comment by xoviat** [comment]
One more interesting bit is that there shouldn't be an `ifdef` here:

https://github.com/flang-compiler/flang/blob/bd6802a4497103debeeed1f4d8f50e3740427cbc/runtime/flang/amod.c#L28

**Comment by xoviat** [comment]
It should be 

```
float
__mth_i_amod(float f, float g)
{
  return FMODF(f, g);
}
```

**Comment by xoviat** [comment]
`sincos`:

```
extern void sincos(double x, double *sine, double *cosine) {
    sine = sin(x);
    cosine = cos(x);
}
```

**Comment by xoviat** [comment]
For `sbrk`, you'll need an `ifdef` here: 

https://github.com/flang-compiler/flang/blob/a9ccdd766d5e606a1dc073fa3b106e52bab5dc64/runtime/flang/xfer_heap_dum.c#L29

Use `VirtualAlloc` for WIN32.

**Comment by xoviat** [comment]
Actually the `sbrk` implementation is a bit trickier. I've opened emeryberger/Malloc-Implementations#1 to attempt to obtain a legally-compatible implementation in a time-efficient way. I currently don't have time to re-implement a memory allocator.

**Comment by isuruf** [comment]
I've pushed some fixes to my branch and the errors now are,
```
[ 69%] Linking Fortran shared library ..\..\bin\flang.dll
LINK: command "C:\PROGRA~2\MICROS~1.0\VC\bin\amd64\link.exe /nologo @CMakeFiles\flang_shared.dir\objects1.rsp /out:..\..\bin\flang.dll /implib:..\..\lib\flang.lib /pdb:C:\Users\idf2\flang\build3\bin\flang.pdb /dll /version:0.0 /machine:x64 /INCREMENTAL:NO /DEF:CMakeFiles\flang_shared.dir\exports.def -LIBPATH:C:\Users\idf2\miniconda3\Library\lib ..\..\lib\flangrti.lib libomp.lib /MANIFEST /MANIFESTFILE:..\..\bin\flang.dll.manifest" failed (exit code 1120) with the following output:
etime3f.c.obj : error LNK2005: convert_filetime already defined in dtime3f.c.obj
etime3f.c.obj : error LNK2005: dtime_ already defined in dtime3f.c.obj
cvt.c.obj : warning LNK4197: export 'ecvt' specified multiple times; using first specification
cvt.c.obj : warning LNK4197: export 'fcvt' specified multiple times; using first specification
   Creating library ..\..\lib\flang.lib and object ..\..\lib\flang.exp
dmod.c.obj : error LNK2019: unresolved external symbol __fmth_i_dmod referenced in function __mth_i_dmod
fsync3f.c.obj : error LNK2019: unresolved external symbol fsync referenced in function fsync_
irandm3f.c.obj : error LNK2019: unresolved external symbol srand48 referenced in function irandm_
irandm3f.c.obj : error LNK2019: unresolved external symbol lrand48 referenced in function irandm_
xfer_heap_dum.c.obj : error LNK2019: unresolved external symbol sbrk referenced in function __fort_sbrk
type.c.obj : error LNK2019: unresolved external symbol __get_fort_size_of referenced in function f90_poly_asn
utils.c.obj : error LNK2019: unresolved external symbol ftruncate referenced in function __fortio_rwinit
init.c.obj : error LNK2019: unresolved external symbol __fort_sethand referenced in function __fort_begpar
..\..\bin\flang.dll : fatal error LNK1120: 8 unresolved externals
```

**Comment by albertziegenhagel** [comment]
FYI: @diorcety has a branch [1] that ported flang to windows in his own fork as well. I don't know how far he got but I think he did very similar changes to yours. Maybe your work can benefit from one another:

[1] https://github.com/diorcety/flang/tree/win32

**Comment by xoviat** [comment]
He doesn't have sbrk either!

**Comment by isuruf** [comment]
That branch is has almost all of our fixes and some not in this PR.

**Comment by xoviat** [comment]
@albertziegenhagel It would be helpful if you wrote functions for some of those symbols.

**Comment by xoviat** [comment]
Looks like no one responded to his first PR and he gave up. That is sad.

**Comment by albertziegenhagel** [comment]
I may have time to look into some of the functions this weekend.

It would be interesting to see how mature the windows files that @sscalpone mentioned are. I don't think it makes sense to do all of the work twice in the end.

**Comment by xoviat** [comment]
Switching the branch...

**Comment by xoviat** [comment]
Continues in #288.

**Pull request closed without merging**